### PR TITLE
Implements focus handling and hover for Material buttons.

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

This implements focus and hover handling for Material buttons. It inserts `Focus` widgets into the tree in order to allow buttons to be focusable via keyboard traversal (a.k.a. TAB traversal), and `Listener` widgets into the `InkWell` to allow the detection of hover states for widgets.

## Related Issues

Addresses #11344, #1608, and #13264.

## Tests

I added the following tests:
 - Tests for `InkWell`
 - Tests for `MaterialButton`
 - Tests for `OutlineButton`, `RaisedButton`, and `FlatButton`

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I signed the [CLA].
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [X] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [X] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/material/theme_data.dart [hunk 1]
```diff
@@ -1026,11 +1051,10 @@ class ThemeData extends Diagnosticable {
...
   int get hashCode {
-    // The hashValues() function supports up to 20 arguments.
-    return hashValues(
-      // Warning: make sure these properties are in the exact same order as in
-      // operator == and in the raw constructor and in the order of fields in
-      // the class and in the lerp() method.
+    // Warning: For the sanity of the reader, please make sure these properties
+    // are in the exact same order as in operator == and in the raw constructor
+    // and in the order of fields in the class and in the lerp() method.
+    final List<Object> values = <Object>[
```
Comment (darrenaustin) [inline packages/flutter/lib/src/material/theme_data.dart]: Why was the original code using the hashValues function instead of the hashList that you are using? Was it a performance concern or somesuch?

File: packages/flutter/test/material/floating_action_button_test.dart [hunk 2]
```diff
@@ -687,20 +704,26 @@ void main() {
...
-          textDirection: TextDirection.ltr,
-          child: Material(
-            child: FloatingActionButton(
-              onPressed: () { /* to make sure the button is enabled */ },
-            ),
+        textDirection: TextDirection.ltr,
+        child: Material(
+          child: FloatingActionButton(
+            focusNode: focusNode,
+            clipBehavior: Clip.none,
```
Comment (darrenaustin) [inline packages/flutter/test/material/floating_action_button_test.dart]: Doesn't this defeat the purpose of this test? It is no longer default if you are forcing it to `Clip.none`.

File: packages/flutter/test/material/ink_well_test.dart [hunk 3]
```diff
@@ -92,6 +90,68 @@ void main() {
...
+      ),
+    ));
+    final TestGesture gesture = await tester.createGesture(kind: PointerDeviceKind.mouse);
+    await gesture.addPointer();
+    await gesture.moveTo(tester.getCenter(find.byType(Container)));
+    await tester.pumpAndSettle();
+    final RenderObject inkFeatures = tester.allRenderObjects.firstWhere((RenderObject object) => object.runtimeType.toString() == '_RenderInkFeatures');
+    expect(inkFeatures, paints..rect(rect: const Rect.fromLTRB(350.0, 250.0, 450.0, 350.0), color: const Color(0xff00ff00)));
+  });
+  testWidgets('ink response changes color on focus', (WidgetTester tester) async {
```
Comment (darrenaustin) [inline packages/flutter/test/material/ink_well_test.dart]: Minor nit: add a blank line above and below this function.

File: packages/flutter/test/material/outline_button_test.dart [hunk 4]
```diff
@@ -236,9 +236,9 @@ void main() {
         child: Material(
           child: Center(
             child: OutlineButton(
-                key: buttonKey,
-                onPressed: () { },
-                child: const Text('ABC'),
+              key: buttonKey,
+              onPressed: () { },
```
Comment (darrenaustin) [inline packages/flutter/test/material/outline_button_test.dart]: Remove space between the {}

File: packages/flutter/test/material/page_test.dart [hunk 5]
```diff
@@ -297,7 +297,7 @@ void main() {
       ),
     );
     await tester.tap(find.text('PUSH'));
-    expect(await tester.pumpAndSettle(const Duration(minutes: 1)), 2);
+    expect(await tester.pumpAndSettle(const Duration(minutes: 1)), 3);
```
Comment (darrenaustin) [inline packages/flutter/test/material/page_test.dart]: What was the reason for this change?

File: packages/flutter/test/material/tabs_test.dart [hunk 6]
```diff
@@ -2229,7 +2229,6 @@ void main() {
       ),
     );
 
-    expect(tester.hasRunningAnimations, isFalse);
```
Comment (darrenaustin) [inline packages/flutter/test/material/tabs_test.dart]: Why was this removed?

**Review by darrenaustin — APPROVED** [review]
Some minor comments and questions aside, LGTM.

File: packages/flutter/lib/src/material/flat_button.dart [hunk 7]
```diff
@@ -163,39 +172,27 @@ class FlatButton extends MaterialButton {
...
       elevation: buttonTheme.getElevation(this),
+      focusElevation: buttonTheme.getFocusElevation(this),
+      hoverElevation: buttonTheme.getHoverElevation(this),
       highlightElevation: buttonTheme.getHighlightElevation(this),
       disabledElevation: buttonTheme.getDisabledElevation(this),
       padding: buttonTheme.getPadding(this),
       constraints: buttonTheme.getConstraints(this),
       shape: buttonTheme.getShape(this),
-      animationDuration: buttonTheme.getAnimationDuration(this),
+      clipBehavior: clipBehavior ?? Clip.none,
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply packages/flutter/lib/src/material/flat_button.dart]: Maybe I'll do both?  That way it's more self-documenting, but if a null gets through, it'll still default to none.

File: packages/flutter/lib/src/material/flat_button.dart [hunk 8]
```diff
@@ -163,39 +172,27 @@ class FlatButton extends MaterialButton {
...
+      focusNode: focusNode,
       materialTapTargetSize: buttonTheme.getMaterialTapTargetSize(this),
+      animationDuration: buttonTheme.getAnimationDuration(this),
       child: child,
     );
   }
-
-  @override
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply packages/flutter/lib/src/material/flat_button.dart]: Yeah, I moved these all up to `MaterialButton`, from all of the `MaterialButton` subclasses, since they all come from there anyhow, and the output was inconsistent between subclasses.  The subclasses should only have things that are additional over `MaterialButton`.

File: packages/flutter/lib/src/material/floating_action_button.dart [hunk 9]
```diff
@@ -126,17 +127,24 @@ class FloatingActionButton extends StatelessWidget {
...
     this.elevation,
+    this.focusElevation,
+    this.hoverElevation,
     this.highlightElevation,
     this.disabledElevation,
     @required this.onPressed,
     this.mini = false,
     this.shape,
-    this.clipBehavior = Clip.none,
+    this.clipBehavior,
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply packages/flutter/lib/src/material/floating_action_button.dart]: It doesn't change the default, just made it so that the `MaterialButton` was where the default was defined (since passing `null` selects the default of `Clip.none`).

However, I think I'll do both: select `Clip.none` if it's `null`, and also show it as the default as a form of documentation.

File: packages/flutter/lib/src/material/ink_well.dart [hunk 10]
```diff
@@ -409,47 +455,110 @@ class InkResponse extends StatefulWidget {
...
+  focus,
+}
+
 class _InkResponseState<T extends InkResponse> extends State<T> with AutomaticKeepAliveClientMixin<T> {
   Set<InteractiveInkFeature> _splashes;
   InteractiveInkFeature _currentSplash;
-  InkHighlight _lastHighlight;
+  FocusNode _focusNode;
+  final Map<_HighlightType, InkHighlight> _highlights = <_HighlightType, InkHighlight>{};
+
```
[the remark this answers is not in the document]
Comment (gspencergoog) [reply packages/flutter/lib/src/material/ink_well.dart]: Yes.  You can have a hover, a focus, and a highlight at the same time. (Click on a button that is focused, for instance, since the mouse is over the button, it's got focus, and it's being pressed), and they can't be the same highlight and just change colors, since they are also all animated.

File: packages/flutter/lib/src/material/theme_data.dart (hunk 1 above)
Comment (gspencergoog) [reply packages/flutter/lib/src/material/theme_data.dart]: No, it's not that,`hashList` just didn't exist when this `hashCode` function was first added.

File: packages/flutter/test/material/floating_action_button_test.dart (hunk 2 above)
Comment (gspencergoog) [reply packages/flutter/test/material/floating_action_button_test.dart]: Duh.  I must have been testing a theory... Removed.

File: packages/flutter/test/material/ink_well_test.dart (hunk 3 above)
Comment (gspencergoog) [reply packages/flutter/test/material/ink_well_test.dart]: Done.

File: packages/flutter/test/material/page_test.dart (hunk 5 above)
Comment (gspencergoog) [reply packages/flutter/test/material/page_test.dart]: Because the hover support adds another frame to the animation.

File: packages/flutter/test/material/tabs_test.dart (hunk 6 above)
Comment (gspencergoog) [reply packages/flutter/test/material/tabs_test.dart]: Actually, it doesn't need to be removed.  For a while I had a bug (but didn't know it) that had the hover animation still running here.  It's fixed now.

File: packages/flutter/test/material/outline_button_test.dart (hunk 4 above)
Comment (gspencergoog) [reply packages/flutter/test/material/outline_button_test.dart]: Done, and elsewhere in this file.  I'm not sure, but I think Hans prefers the extra space, however.
